contrib/hashicorp/vault/vault_test.go

```diff
@@ -0,0 +1,442 @@
+		if err != nil {
+			t.Fatal(err)
+		}
+		spans := mt.FinishedSpans()
+		assert.Len(spans, 1)
+		span := spans[0]
+
+		assert.Equal("vault", span.Tag(ext.ServiceName))
+		assert.Equal(fullPath, span.Tag(ext.HTTPURL))
+		assert.Equal(http.MethodPut, span.Tag(ext.HTTPMethod))
+		assert.Equal(http.MethodPut+" "+fullPath, span.Tag(ext.ResourceName))
+		assert.Equal(ext.SpanTypeHTTP, span.Tag(ext.SpanType))
+		assert.Equal(200, span.Tag(ext.HTTPCode))
+		assert.Zero(span.Tag(ext.Error))
+		assert.Zero(span.Tag(ext.ErrorMsg))
+		assert.Equal(namespace, span.Tag("vault.namespace"))
+	})
+
+	t.Run("read", func(t *testing.T) {
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		// Write key with namespace first
+		data := map[string]interface{}{"Key1": "Val1", "Key2": "Val2"}
+		_, err = client.Logical().Write(key, data)
+		if err != nil {
+			t.Fatal(err)
+		}
+		// Read key with namespace
+		_, err = client.Logical().Read(key)
+		if err != nil {
+			t.Fatal(err)
+		}
+		spans := mt.FinishedSpans()
+		assert.Len(spans, 2)
+		span := spans[1]
+
+		assert.Equal("vault", span.Tag(ext.ServiceName))
+		assert.Equal(fullPath, span.Tag(ext.HTTPURL))
+		assert.Equal(http.MethodGet, span.Tag(ext.HTTPMethod))
+		assert.Equal(http.MethodGet+" "+fullPath, span.Tag(ext.ResourceName))
+		assert.Equal(ext.SpanTypeHTTP, span.Tag(ext.SpanType))
+		assert.Equal(200, span.Tag(ext.HTTPCode))
+		assert.Zero(span.Tag(ext.Error))
+		assert.Zero(span.Tag(ext.ErrorMsg))
+		assert.Equal(namespace, span.Tag("vault.namespace"))
+	})
+}
+
+func getWriteSpan(t *testing.T, conf *api.Config) mocktracer.Span {
+	assert := assert.New(t)
+	client, err := api.NewClient(conf)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer mountKV(client, t)()
+
+	mt := mocktracer.Start()
+	defer mt.Stop()
+
+	_, err = client.Logical().Write(
+		secretMountPath+"/key",
+		map[string]interface{}{"Key1": "Val1", "Key2": "Val2"},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	spans := mt.FinishedSpans()
+	assert.Len(spans, 1)
+	return spans[0]
+}
+
+func TestOptionServiceName(t *testing.T) {
+	assert := assert.New(t)
+
+	ts, cleanupTs := setupServer(t)
+	defer cleanupTs()
+
+	t.Run("WithServiceName", func(t *testing.T) {
+		// Check default
+		config := &api.Config{
+			HttpClient: NewHTTPClient(),
+			Address:    ts.URL,
+		}
+		span := getWriteSpan(t, config)
+		assert.Equal(serviceName, span.Tag(ext.ServiceName))
+
+		customServiceName := "someServiceName"
+		config = &api.Config{
+			HttpClient: NewHTTPClient(WithServiceName(customServiceName)),
+			Address:    ts.URL,
+		}
+		span = getWriteSpan(t, config)
+		assert.Equal(customServiceName, span.Tag(ext.ServiceName))
+	})
+
+	t.Run("WithAnalytics", func(t *testing.T) {
+		// With no option, default should be no tag.
+		config := &api.Config{
+			HttpClient: NewHTTPClient(),
+			Address:    ts.URL,
+		}
+		span := getWriteSpan(t, config)
+		// Zero value for type (nil), not 0.0 float value
+		assert.Zero(span.Tag(ext.EventSampleRate))
```

Comment:
@gbbr  I repeated this comment everywhere for `assert.Zero` of the `EventSampleRate` because I find it confusing, but maybe it's overkill. Let me know what you think.

Reply:
If we have to repeat a comment above every line of code, it's a sign that something is wrong somewhere. The documentation for `assert.Zero` should clarify what it does for anybody. Lastly, you can use `assert.Nil` here if that's what you're after.